// @yext/search-ui-react/lib/esm/components/cards/standard/StandardCard.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useCardAnalyticsCallback } from '../../../hooks/useCardAnalyticsCallback';
import { useCardFeedbackCallback } from '../../../hooks/useCardFeedbackCallback';
import { StandardCardDisplay } from './StandardCardDisplay';
/**
 * This Component renders the base result card.
 *
 * @public
 *
 * @param props - An object containing the result itself and any additional information needed
 *                to render the card
 * @returns A React element for the result card
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function StandardCard(props) {
    var _a, _b, _c, _d, _e;
    const { result, customCssClasses, showFeedbackButtons } = props;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const data = {
        title: (_c = (_b = (_a = result.highlightedFields) === null || _a === void 0 ? void 0 : _a.name) !== null && _b !== void 0 ? _b : result.name) !== null && _c !== void 0 ? _c : result.rawData.name,
        description: (_e = (_d = result.highlightedFields) === null || _d === void 0 ? void 0 : _d.description) !== null && _e !== void 0 ? _e : result.rawData.description,
        cta1: result.rawData.c_primaryCTA,
        cta2: result.rawData.c_secondaryCTA,
    };
    // eslint-disable-next-line react-perf/jsx-no-new-object-as-prop
    const clickHandlers = {
        handleCtaClick: useCardAnalyticsCallback(result, 'CTA_CLICK'),
        handleTitleClick: useCardAnalyticsCallback(result, 'TITLE_CLICK'),
        handleFeedbackClick: useCardFeedbackCallback(result)
    };
    /*
      Note the title, link, description, and CTA props are sourced from LiveAPI data. This data
      must be type-checked at run-time. This is accomplished by using the StandardCardDisplay's PropTypes.
    */
    return (_jsx(StandardCardDisplay, { customCssClasses: customCssClasses, showFeedbackButtons: showFeedbackButtons, clickHandlers: clickHandlers, title: data.title, link: result.link, description: data.description, cta1: data.cta1, cta2: data.cta2 }, void 0));
}
