// react-cmdk/dist/components/ListItem.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Button = Button;
exports.Link = Link;
exports["default"] = ListItem;

var _Icon = _interopRequireDefault(require("./Icon"));

var _react = _interopRequireWildcard(require("react"));

var _context = require("../lib/context");

var _utils = require("../lib/utils");

var _excluded = ["renderLink", "closeOnSelect", "disabled", "showType", "className", "iconType", "children", "onClick", "index", "icon"],
    _excluded2 = ["closeOnSelect", "showType", "className", "children", "iconType", "onClick", "index", "icon"];

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function getListItemWrapperStyles(selected, disabled) {
  return (0, _utils.classNames)("command-palette-list-item block w-full text-left px-3.5 py-2.5 rounded-md hover:bg-gray-100 dark:hover:bg-gray-800 focus:ring-1 focus:ring-gray-300 focus:outline-none flex items-center space-x-2.5 justify-between", selected && !disabled ? "bg-gray-200/50 dark:bg-gray-800" : "bg-transparent", disabled ? "cursor-default pointer-events-none opacity-50" : "cursor-pointer");
}

function Link(_ref) {
  var localRenderLink = _ref.renderLink,
      _ref$closeOnSelect = _ref.closeOnSelect,
      closeOnSelect = _ref$closeOnSelect === void 0 ? true : _ref$closeOnSelect,
      _ref$disabled = _ref.disabled,
      disabled = _ref$disabled === void 0 ? false : _ref$disabled,
      _ref$showType = _ref.showType,
      showType = _ref$showType === void 0 ? true : _ref$showType,
      className = _ref.className,
      iconType = _ref.iconType,
      children = _ref.children,
      onClick = _ref.onClick,
      index = _ref.index,
      icon = _ref.icon,
      rest = _objectWithoutProperties(_ref, _excluded);

  var _useContext = (0, _react.useContext)(_context.RenderLinkContext),
      globalRenderLink = _useContext.renderLink;

  var _useContext2 = (0, _react.useContext)(_context.OpenContext),
      onChangeOpen = _useContext2.onChangeOpen;

  var _useContext3 = (0, _react.useContext)(_context.SelectContext),
      selected = _useContext3.selected;

  var renderLink = localRenderLink || globalRenderLink;

  function renderLinkContent() {
    return /*#__PURE__*/_react["default"].createElement(ListItemContent, {
      type: showType ? "Link" : undefined,
      iconType: iconType,
      icon: icon
    }, children);
  }

  var styles = (0, _utils.classNames)(getListItemWrapperStyles(selected === index, disabled), className);

  function clickAndClose(e) {
    if (rest.href && !disabled) {
      if (onClick) {
        onClick(e);
      }

      if (closeOnSelect) {
        onChangeOpen(false);
      }
    }
  }

  return renderLink ? /*#__PURE__*/_react["default"].createElement(_react.Fragment, null, renderLink(_objectSpread(_objectSpread({}, rest), {}, {
    "data-close-on-select": closeOnSelect,
    children: renderLinkContent(),
    "aria-disabled": disabled,
    onClick: clickAndClose,
    className: styles
  }))) : /*#__PURE__*/_react["default"].createElement("a", _extends({}, rest, {
    "data-close-on-select": closeOnSelect,
    "aria-disabled": disabled,
    onClick: clickAndClose,
    className: styles
  }), renderLinkContent());
}

function Button(_ref2) {
  var _rest$disabled;

  var _ref2$closeOnSelect = _ref2.closeOnSelect,
      closeOnSelect = _ref2$closeOnSelect === void 0 ? true : _ref2$closeOnSelect,
      _ref2$showType = _ref2.showType,
      showType = _ref2$showType === void 0 ? true : _ref2$showType,
      className = _ref2.className,
      children = _ref2.children,
      iconType = _ref2.iconType,
      onClick = _ref2.onClick,
      index = _ref2.index,
      icon = _ref2.icon,
      rest = _objectWithoutProperties(_ref2, _excluded2);

  var _useContext4 = (0, _react.useContext)(_context.SelectContext),
      selected = _useContext4.selected;

  var _useContext5 = (0, _react.useContext)(_context.OpenContext),
      onChangeOpen = _useContext5.onChangeOpen;

  function clickAndClose(e) {
    if (onClick) {
      onClick(e);

      if (closeOnSelect) {
        onChangeOpen(false);
      }
    }
  }

  return /*#__PURE__*/_react["default"].createElement("button", _extends({}, rest, {
    "aria-disabled": (_rest$disabled = rest.disabled) !== null && _rest$disabled !== void 0 ? _rest$disabled : false,
    "data-close-on-select": closeOnSelect,
    onClick: clickAndClose,
    className: (0, _utils.classNames)(getListItemWrapperStyles(selected === index, rest.disabled), className)
  }), /*#__PURE__*/_react["default"].createElement(ListItemContent, {
    type: showType ? "Action" : undefined,
    iconType: iconType,
    icon: icon
  }, children));
}

function ListItemContent(_ref3) {
  var ListItemIcon = _ref3.icon,
      iconType = _ref3.iconType,
      children = _ref3.children,
      type = _ref3.type;
  return /*#__PURE__*/_react["default"].createElement(_react["default"].Fragment, null, /*#__PURE__*/_react["default"].createElement("div", {
    className: "flex w-full items-center space-x-2.5"
  }, ListItemIcon && (typeof ListItemIcon === "string" ? /*#__PURE__*/_react["default"].createElement(_Icon["default"], {
    name: ListItemIcon,
    type: iconType
  }) : /*#__PURE__*/_react["default"].createElement(ListItemIcon, {
    className: "w-5 h-5 text-gray-500"
  })), typeof children === "string" ? /*#__PURE__*/_react["default"].createElement("span", {
    className: "truncate max-w-md dark:text-white"
  }, children) : children), type && /*#__PURE__*/_react["default"].createElement("span", {
    className: "text-gray-500 text-sm"
  }, type));
}

function ListItem(props) {
  var Wrapper = props.href ? Link : Button;
  return /*#__PURE__*/_react["default"].createElement(Wrapper, props);
}
