// react-cmdk/dist/components/Search.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireWildcard(require("react"));

var _outline = require("@heroicons/react/outline");

var _solid = require("@heroicons/react/solid");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function Search(_ref, ref) {
  var _onChange = _ref.onChange,
      placeholder = _ref.placeholder,
      prefix = _ref.prefix,
      value = _ref.value;
  return /*#__PURE__*/_react["default"].createElement("div", {
    className: "flex items-center space-x-1.5 pl-3"
  }, /*#__PURE__*/_react["default"].createElement(_outline.SearchIcon, {
    className: "w-4 pointer-events-none text-gray-400 dark:text-gray-600"
  }), prefix !== null && prefix !== void 0 && prefix.length ? prefix.map(function (p) {
    return /*#__PURE__*/_react["default"].createElement(_react.Fragment, {
      key: p
    }, /*#__PURE__*/_react["default"].createElement("span", {
      className: "dark:text-white"
    }, p), /*#__PURE__*/_react["default"].createElement("span", {
      className: "text-gray-500"
    }, "/"));
  }) : null, /*#__PURE__*/_react["default"].createElement("div", {
    className: "flex-1 relative"
  }, /*#__PURE__*/_react["default"].createElement("input", {
    ref: ref,
    spellCheck: false,
    className: "py-4 px-0 border-none w-full focus:outline-none focus:border-none focus:ring-0 bg-transparent placeholder-gray-500 dark:text-white",
    onChange: function onChange(e) {
      _onChange(e.currentTarget.value);
    },
    onFocus: function onFocus(e) {
      e.currentTarget.select();
    },
    onKeyDown: function onKeyDown(e) {
      if (e.key === "Escape" && value) {
        e.preventDefault();
        e.stopPropagation();

        _onChange("");
      }
    },
    id: "command-palette-search-input",
    placeholder: placeholder,
    value: value,
    type: "text",
    autoFocus: true
  }), value && /*#__PURE__*/_react["default"].createElement("button", {
    tabIndex: -1,
    type: "button",
    onClick: function onClick() {
      _onChange("");

      var inputElement = document.getElementById("command-palette-search-input");

      if (inputElement) {
        inputElement.focus();
      }
    }
  }, /*#__PURE__*/_react["default"].createElement(_solid.XCircleIcon, {
    className: "w-5 text-gray-300 dark:text-gray-600 hover:text-gray-500 dark:hover:text-gray-300 transition absolute right-3 top-1/2 transform -translate-y-1/2"
  }))));
}

var _default = /*#__PURE__*/(0, _react.forwardRef)(Search);

exports["default"] = _default;
