// @yext/pages/dist/vite-plugin/build/closeBundle/manifest.js
import fs from "fs-extra";
const generateManifestFile = (featureNameToBundlePath, projectStructure) => {
  const distRoot = projectStructure.distRoot.getAbsolutePath();
  const relativeBundlePaths = Array.from(featureNameToBundlePath.entries()).map(([name, path]) => [name, projectStructure.distRoot.getRelativePath(path)]);
  let bundlerManifest = Buffer.from("{}");
  if (fs.existsSync(`${distRoot}/manifest.json`)) {
    bundlerManifest = fs.readFileSync(`${distRoot}/manifest.json`);
  }
  const manifest = {
    bundlePaths: Object.fromEntries(relativeBundlePaths),
    projectFilepaths: {
      templatesRoot: projectStructure.templatesRoot.path,
      distRoot: projectStructure.distRoot.path,
      hydrationBundleOutputRoot: projectStructure.hydrationBundleOutputRoot.path,
      serverBundleOutputRoot: projectStructure.serverBundleOutputRoot.path
    },
    bundlerManifest: JSON.parse(bundlerManifest.toString())
  };
  fs.writeFileSync(`${distRoot}/plugin/manifest.json`, JSON.stringify(manifest, null, "  "));
  fs.remove(`${distRoot}/manifest.json`);
};
export {
  generateManifestFile
};
