// @yext/search-ui-react/lib/commonjs/components/AlternativeVerticals.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlternativeVerticals = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const processTranslation_1 = require("./utils/processTranslation");
const StarIcon_1 = require("../icons/StarIcon");
const search_headless_react_1 = require("@yext/search-headless-react");
const useComposedCssClasses_1 = require("../hooks/useComposedCssClasses");
const classnames_1 = __importDefault(require("classnames"));
const builtInCssClasses = {
    alternativeVerticalsContainer: 'flex flex-col justify-between border rounded-lg mb-4 p-4 shadow-sm',
    alternativeVerticalsLoading: 'opacity-50',
    noResultsText: 'text-lg text-neutral-dark pb-2',
    categoriesText: 'text-neutral',
    suggestion: 'pb-4 flex items-center',
    verticalIcon: 'w-4 mr-2'
};
function isVerticalSuggestion(suggestion) {
    var _a, _b;
    return ((_a = suggestion) === null || _a === void 0 ? void 0 : _a.resultsCount) !== undefined &&
        ((_b = suggestion) === null || _b === void 0 ? void 0 : _b.verticalKey) !== undefined;
}
/**
 * A component that displays the alternative verticals that have results if a search returns none
 * on the current vertical.
 *
 * @public
 *
 * @param props - {@link AlternativeVerticalsProps}
 * @returns A React element for the alternative verticals, or null if there are none with results
 */
function AlternativeVerticals({ currentVerticalLabel, verticalConfigMap, displayAllOnNoResults = true, customCssClasses }) {
    var _a;
    const cssClasses = (0, useComposedCssClasses_1.useComposedCssClasses)(builtInCssClasses, customCssClasses);
    const alternativeVerticals = (0, search_headless_react_1.useSearchState)(state => { var _a; return (_a = state.vertical.noResults) === null || _a === void 0 ? void 0 : _a.alternativeVerticals; }) || [];
    const allResultsForVertical = (0, search_headless_react_1.useSearchState)(state => { var _a; return (_a = state.vertical.noResults) === null || _a === void 0 ? void 0 : _a.allResultsForVertical.results; }) || [];
    const query = (0, search_headless_react_1.useSearchState)(state => state.query.mostRecentSearch);
    const verticalSuggestions = buildVerticalSuggestions(verticalConfigMap, alternativeVerticals);
    const isShowingAllResults = displayAllOnNoResults && allResultsForVertical.length > 0;
    const isLoading = (0, search_headless_react_1.useSearchState)(state => state.searchStatus.isLoading);
    const containerClassNames = (0, classnames_1.default)(cssClasses.alternativeVerticalsContainer, {
        [(_a = cssClasses.alternativeVerticalsLoading) !== null && _a !== void 0 ? _a : '']: isLoading
    });
    function buildVerticalSuggestions(verticalConfigMap, alternativeVerticals) {
        return alternativeVerticals
            .filter((alternativeResults) => {
            return !!verticalConfigMap[alternativeResults.verticalKey];
        })
            .map((alternativeResults) => {
            return {
                label: verticalConfigMap[alternativeResults.verticalKey].label,
                verticalKey: alternativeResults.verticalKey,
                resultsCount: alternativeResults.resultsCount
            };
        })
            .filter(isVerticalSuggestion)
            .filter(verticalSuggestion => verticalSuggestion.resultsCount > 0);
    }
    if (verticalSuggestions.length <= 0) {
        return null;
    }
    return ((0, jsx_runtime_1.jsxs)("div", Object.assign({ className: containerClassNames }, { children: [renderNoResultsInfo(), verticalSuggestions &&
                (0, jsx_runtime_1.jsxs)("div", Object.assign({ className: 'pt-4 text-neutral-dark' }, { children: [(0, jsx_runtime_1.jsxs)("div", Object.assign({ className: cssClasses.categoriesText }, { children: [(0, jsx_runtime_1.jsx)("span", { children: (0, processTranslation_1.processTranslation)({
                                        phrase: 'The following category yielded results for - ',
                                        pluralForm: 'The following categories yielded results for - ',
                                        count: verticalSuggestions.length
                                    }) }, void 0), (0, jsx_runtime_1.jsx)("strong", { children: query }, void 0)] }), void 0), (0, jsx_runtime_1.jsx)("ul", Object.assign({ className: 'pt-4' }, { children: verticalSuggestions.map(renderSuggestion) }), void 0)] }), void 0)] }), void 0));
    function renderNoResultsInfo() {
        return ((0, jsx_runtime_1.jsxs)("div", Object.assign({ className: cssClasses.noResultsText }, { children: [(0, jsx_runtime_1.jsxs)("span", { children: ["No results found in ", currentVerticalLabel, "."] }, void 0), isShowingAllResults &&
                    (0, jsx_runtime_1.jsxs)("span", { children: [" Showing all ", currentVerticalLabel, " instead."] }, void 0)] }), void 0));
    }
    function renderSuggestion(suggestion) {
        const resultsCountText = (0, processTranslation_1.processTranslation)({
            phrase: `${suggestion.resultsCount} result`,
            pluralForm: `${suggestion.resultsCount} results`,
            count: suggestion.resultsCount
        });
        return ((0, jsx_runtime_1.jsxs)("li", Object.assign({ className: cssClasses.suggestion }, { children: [(0, jsx_runtime_1.jsx)("div", Object.assign({ className: cssClasses.verticalIcon }, { children: (0, jsx_runtime_1.jsx)(StarIcon_1.StarIcon, {}, void 0) }), void 0), (0, jsx_runtime_1.jsxs)("span", Object.assign({ className: 'font-bold' }, { children: [suggestion.label, " - ", resultsCountText] }), void 0)] }), suggestion.verticalKey));
    }
}
exports.AlternativeVerticals = AlternativeVerticals;
