// @yext/search-ui-react/lib/commonjs/components/Filters/FacetsProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FacetsProvider = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const search_headless_react_1 = require("@yext/search-headless-react");
const react_1 = require("react");
const filterutils_1 = require("../../utils/filterutils");
const filterutils_2 = require("../../utils/filterutils");
const search_operations_1 = require("../../utils/search-operations");
const FiltersContext_1 = require("./FiltersContext");
/**
 * The Facets component is a wrapper component around {@link Filters} that updates facet options
 * when a child filter is updated.
 *
 * The representation of the facets is configured using a FACC (function as a child component)
 * The FACC is passed the facets data, and is intended for use with components like
 * {@link Filters.CheckboxOption}.
 *
 * @public
 *
 * @param props - {@link Filters.FacetsProviderProps}
 */
function FacetsProvider({ children, className = 'w-full', searchOnChange = true }) {
    const searchActions = (0, search_headless_react_1.useSearchActions)();
    const facetsInState = (0, search_headless_react_1.useSearchState)(state => state.filters.facets);
    const facets = (0, react_1.useMemo)(() => facetsInState !== null && facetsInState !== void 0 ? facetsInState : [], [facetsInState]);
    const filters = (0, react_1.useMemo)(() => {
        return facets.flatMap(f => f.options.map(o => {
            return {
                fieldId: f.fieldId,
                value: o.value,
                matcher: o.matcher,
                selected: o.selected,
                displayName: o.displayName
            };
        }));
    }, [facets]);
    const filtersContextInstance = (0, react_1.useMemo)(() => {
        return {
            selectFilter(filter) {
                if (typeof filter.value === 'object' && !(0, filterutils_1.isNumberRangeValue)(filter.value)) {
                    console.error('Facets only support string, number, boolean, and NumberRangeValue. Found the following object value instead:', filter.value);
                    return;
                }
                const facetOption = {
                    matcher: filter.matcher,
                    value: filter.value
                };
                searchActions.setFacetOption(filter.fieldId, facetOption, filter.selected);
            },
            applyFilters() {
                if (searchOnChange) {
                    searchActions.setOffset(0);
                    (0, filterutils_2.clearStaticRangeFilters)(searchActions, (0, filterutils_1.getSelectedNumericalFacetFields)(searchActions));
                    (0, search_operations_1.executeSearch)(searchActions);
                }
            },
            filters
        };
    }, [searchActions, filters, searchOnChange]);
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({ className: className }, { children: (0, jsx_runtime_1.jsx)(FiltersContext_1.FiltersContext.Provider, Object.assign({ value: filtersContextInstance }, { children: children === null || children === void 0 ? void 0 : children(facets) }), void 0) }), void 0));
}
exports.FacetsProvider = FacetsProvider;
