// @yext/search-ui-react/lib/commonjs/components/Filters/HierarchicalFacetDisplay.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HierarchicalFacetDisplay = exports.DEFAULT_HIERARCHICAL_DELIMITER = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const useComposedCssClasses_1 = require("../../hooks/useComposedCssClasses");
const useHierarchicalFacetTree_1 = require("../../hooks/useHierarchicalFacetTree");
const FiltersContext_1 = require("./FiltersContext");
const builtInCssClasses = {
    treeContainer: 'flex flex-col items-start',
    allCategoriesOption___active: 'font-semibold mb-2 text-sm',
    allCategoriesOption___inactive: 'mb-2 text-sm',
    availableOption__active: 'font-semibold ml-4 mb-2 text-sm',
    availableOption__inactive: 'ml-4 mb-2 text-sm',
    parentCategory: 'mb-2 text-sm',
    currentCategory: 'font-semibold mb-2 text-sm',
    showMoreButton: 'ml-4 text-sm font-medium text-primary'
};
exports.DEFAULT_HIERARCHICAL_DELIMITER = '>';
/**
 * A HierarchicalFacetDisplay takes a `DisplayableFacet` and renders the facet in a way
 * to represent multiple levels of "hierarchies".
 *
 * The hierarchies are determined by the provided delimiter, which defaults to "\>".
 *
 * @public
 */
function HierarchicalFacetDisplay({ facet, delimiter = exports.DEFAULT_HIERARCHICAL_DELIMITER, showMoreLimit = 4, customCssClasses }) {
    const cssClasses = (0, useComposedCssClasses_1.useComposedCssClasses)(builtInCssClasses, customCssClasses);
    const tree = (0, useHierarchicalFacetTree_1.useHierarchicalFacetTree)(facet, delimiter);
    const [isShowingMore, setIsShowingMore] = (0, react_1.useState)(false);
    const resetShowMore = (0, react_1.useCallback)(() => setIsShowingMore(false), []);
    const toggleShowMore = (0, react_1.useCallback)(() => {
        setIsShowingMore(!isShowingMore);
    }, [isShowingMore]);
    /** Iteratively parses the `HierarchicalFacetTree` into an array of ReactNodes */
    function renderTree() {
        let treePointer = tree;
        const renderedNodesAndShowMoreButton = [renderAllCategoriesButton()];
        while (treePointer) {
            const currentNodes = Object.values(treePointer);
            const selectedChildNode = currentNodes.find(n => n.selected);
            const selectedHasNoChildren = selectedChildNode && Object.values(selectedChildNode.childTree).length === 0;
            const activeParentNode = currentNodes.find(n => n.hasSelectedChild);
            if ((!selectedChildNode && !activeParentNode) || selectedHasNoChildren) {
                renderedNodesAndShowMoreButton.push(...renderAvailableOptions(currentNodes));
                if (currentNodes.length > showMoreLimit) {
                    renderedNodesAndShowMoreButton.push(renderShowMoreButton());
                }
                break;
            }
            const activeNode = selectedChildNode !== null && selectedChildNode !== void 0 ? selectedChildNode : activeParentNode;
            if (!activeNode) {
                break;
            }
            renderedNodesAndShowMoreButton.push(renderCategory(activeNode, facet.fieldId));
            treePointer = activeNode.childTree;
        }
        return renderedNodesAndShowMoreButton;
    }
    function renderAllCategoriesButton() {
        return ((0, jsx_runtime_1.jsx)(AllCategories, { activeClassName: cssClasses.allCategoriesOption___active, inactiveClassName: cssClasses.allCategoriesOption___inactive, facet: facet, resetShowMore: resetShowMore }, '_AllCategories'));
    }
    function renderAvailableOptions(nodes) {
        const nodesToRender = isShowingMore ? nodes : nodes.slice(0, showMoreLimit);
        return nodesToRender.map(n => (0, jsx_runtime_1.jsx)(AvailableOption, { activeClassName: cssClasses.availableOption__active, inactiveClassName: cssClasses.availableOption__inactive, fieldId: facet.fieldId, currentNode: n, resetShowMore: resetShowMore, siblingNodes: nodes.filter(siblingNode => siblingNode !== n) }, n.lastDisplayNameToken));
    }
    function renderShowMoreButton() {
        return (0, jsx_runtime_1.jsx)(ShowMoreButton, { className: cssClasses.showMoreButton, isShowingMore: isShowingMore, toggleShowMore: toggleShowMore }, '_ShowMoreButton');
    }
    function renderCategory(selectedNode, fieldId) {
        const sharedProps = {
            key: selectedNode.lastDisplayNameToken,
            resetShowMore,
            selectedNode,
            fieldId
        };
        if (selectedNode.hasSelectedChild) {
            return (0, jsx_runtime_1.jsx)(ParentCategory, Object.assign({ className: cssClasses.parentCategory }, sharedProps), void 0);
        }
        else {
            return (0, jsx_runtime_1.jsx)(CurrentCategory, Object.assign({ className: cssClasses.currentCategory }, sharedProps), void 0);
        }
    }
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({ className: cssClasses.treeContainer }, { children: renderTree() }), void 0));
}
exports.HierarchicalFacetDisplay = HierarchicalFacetDisplay;
/**
 * A hard-coded "All Categories" button that can be used to reset the HierarchicalFacetDisplay
 * to its initial state, i.e. with no options selected.
 */
function AllCategories({ facet, inactiveClassName, activeClassName, resetShowMore }) {
    const { applyFilters, selectFilter } = (0, FiltersContext_1.useFiltersContext)();
    const handleClickAllCategories = (0, react_1.useCallback)(() => {
        facet.options
            .filter(o => o.selected)
            .forEach(o => selectFilter(Object.assign(Object.assign({}, o), { fieldId: facet.fieldId, selected: false })));
        applyFilters();
        resetShowMore();
    }, [applyFilters, facet.fieldId, facet.options, resetShowMore, selectFilter]);
    if (facet.options.find(o => o.selected)) {
        return ((0, jsx_runtime_1.jsx)("button", Object.assign({ className: inactiveClassName, onClick: handleClickAllCategories }, { children: "All Categories /" }), void 0));
    }
    return ((0, jsx_runtime_1.jsx)("div", Object.assign({ className: activeClassName }, { children: "All Categories" }), void 0));
}
/** An option currently available for selection or deselection. */
function AvailableOption(props) {
    var _a;
    const { fieldId, currentNode, activeClassName, inactiveClassName, resetShowMore, siblingNodes } = props;
    const { applyFilters, selectFilter } = (0, FiltersContext_1.useFiltersContext)();
    const { selected, lastDisplayNameToken, facetOption } = currentNode;
    const handleClickAvailableOptions = (0, react_1.useCallback)(() => {
        var _a;
        siblingNodes.filter(n => n.selected).forEach(n => selectFilter(Object.assign(Object.assign({}, n.facetOption), { selected: false, fieldId })));
        selectFilter(Object.assign(Object.assign({}, facetOption), { selected: !selected, fieldId }));
        const parentFacetOption = (_a = currentNode.parentNode) === null || _a === void 0 ? void 0 : _a.facetOption;
        parentFacetOption && selectFilter(Object.assign(Object.assign({}, parentFacetOption), { selected,
            fieldId }));
        applyFilters();
        resetShowMore();
    }, [
        applyFilters,
        (_a = currentNode.parentNode) === null || _a === void 0 ? void 0 : _a.facetOption,
        facetOption,
        fieldId,
        resetShowMore,
        selectFilter,
        selected,
        siblingNodes
    ]);
    return ((0, jsx_runtime_1.jsx)("button", Object.assign({ className: selected ? activeClassName : inactiveClassName, onClick: handleClickAvailableOptions }, { children: lastDisplayNameToken }), void 0));
}
/** A parent category that is not the most immediate one. */
function ParentCategory({ fieldId, selectedNode, className, resetShowMore }) {
    const { applyFilters, selectFilter } = (0, FiltersContext_1.useFiltersContext)();
    const deselectChildOptions = (0, react_1.useCallback)((node) => {
        const tree = node.childTree;
        Object.values(tree).forEach(n => {
            selectFilter(Object.assign(Object.assign({}, n.facetOption), { selected: false, fieldId }));
            deselectChildOptions(n);
        });
    }, [fieldId, selectFilter]);
    const handleClickParentCategory = (0, react_1.useCallback)(() => {
        selectFilter(Object.assign(Object.assign({}, selectedNode.facetOption), { selected: true, fieldId }));
        deselectChildOptions(selectedNode);
        applyFilters();
        resetShowMore();
    }, [applyFilters, deselectChildOptions, fieldId, resetShowMore, selectFilter, selectedNode]);
    return ((0, jsx_runtime_1.jsx)("button", Object.assign({ className: className, onClick: handleClickParentCategory }, { children: selectedNode.lastDisplayNameToken + ' /' }), void 0));
}
/** The currently selected category, i.e. the most immediate category. */
function CurrentCategory({ fieldId, selectedNode, className, resetShowMore }) {
    var _a;
    const { applyFilters, selectFilter } = (0, FiltersContext_1.useFiltersContext)();
    const handleClickCurrentCategory = (0, react_1.useCallback)(() => {
        var _a;
        selectFilter(Object.assign(Object.assign({}, selectedNode.facetOption), { selected: false, fieldId }));
        const parentFacetOption = (_a = selectedNode.parentNode) === null || _a === void 0 ? void 0 : _a.facetOption;
        parentFacetOption && selectFilter(Object.assign(Object.assign({}, parentFacetOption), { selected: true, fieldId }));
        applyFilters();
        resetShowMore();
    }, [
        applyFilters,
        fieldId,
        resetShowMore,
        selectFilter,
        selectedNode.facetOption,
        (_a = selectedNode.parentNode) === null || _a === void 0 ? void 0 : _a.facetOption
    ]);
    return ((0, jsx_runtime_1.jsx)("button", Object.assign({ className: className, onClick: handleClickCurrentCategory }, { children: selectedNode.lastDisplayNameToken }), void 0));
}
/** The "Show more/less" button for hiding/showing additional `AvailableOption`s */
function ShowMoreButton({ className, isShowingMore, toggleShowMore }) {
    return ((0, jsx_runtime_1.jsx)("button", Object.assign({ className: className, onClick: toggleShowMore }, { children: isShowingMore ? 'Show less' : 'Show more' }), void 0));
}
